atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/DefaultClusterEventService.java

```diff
@@ -354,77 +202,64 @@ public boolean isRunning() {
+  public void event(final ClusterMembershipEvent event) {
+    eventServiceExecutor.execute(
+        () -> {
+          switch (event.type()) {
+            case MEMBER_ADDED:
+              updateRemoteSubscription(event);
+              break;
+            case METADATA_CHANGED:
+              updateRemoteSubscription(event);
+              break;
+            case REACHABILITY_CHANGED:
+              break;
+            case MEMBER_REMOVED:
+              removeAllSubscription(event.subject().id());
+              break;
+            default:
+              LOGGER.warn(
+                  "Unexpected membership event type {} from {}", event.type(), event.subject());
+          }
+        });
   }
 
-  /** Subscriber iterator that iterates subscribers in a loop. */
-  private static class TopicIterator implements Iterator<InternalSubscriptionInfo> {
-    private final AtomicInteger counter = new AtomicInteger();
-    private final InternalSubscriptionInfo[] subscribers;
-
-    TopicIterator(final List<InternalSubscriptionInfo> subscribers) {
-      final List<InternalSubscriptionInfo> filteredSubscribers =
-          subscribers.stream().filter(s -> !s.isTombstone()).collect(Collectors.toList());
-      Collections.reverse(filteredSubscribers);
-      this.subscribers =
-          filteredSubscribers.toArray(new InternalSubscriptionInfo[filteredSubscribers.size()]);
-    }
-
-    @Override
-    public boolean hasNext() {
-      return subscribers.length > 0;
+  private void removeAllSubscription(final MemberId id) {
+    final Set<String> prevSubscriptions = remoteMemberSubscriptions.remove(id);
+    if (prevSubscriptions != null) {
+      prevSubscriptions.forEach(s -> topics.get(s).removeRemoteSubscription(id));
     }
+  }
 
-    @Override
-    public InternalSubscriptionInfo next() {
-      return subscribers[Math.abs(counter.incrementAndGet() % subscribers.length)];
+  private void updateRemoteSubscription(final ClusterMembershipEvent event) {
+    final String topicSubscribedAsString =
+        event.subject().properties().getProperty(SUBSCRIPTION_PROPERTY_NAME);
+    if (topicSubscribedAsString != null) {
+      final Set<String> topicsSubscribed = topicsFromString(topicSubscribedAsString);
+      topicsSubscribed.forEach(
+          topic ->
+              topics
+                  .computeIfAbsent(topic, t -> new InternalTopic(topic))
+                  .addRemoteSubscription(event.subject().id()));
+      remoteMemberSubscriptions.put(event.subject().id(), topicsSubscribed);
+    } else {
+      removeAllSubscription(event.subject().id());
     }
   }
 
   /** Internal subscriber. */
   private static class InternalSubscriber
       implements BiFunction<Address, byte[], CompletableFuture<byte[]>> {
-    private final AtomicInteger counter = new AtomicInteger();
-    private InternalSubscription[] subscriptions = new InternalSubscription[0];
+    private final List<InternalSubscription> subscriptions = new CopyOnWriteArrayList<>();
```

Comment:
Do we really need here a concurrent data structure?

Reply:
Yes. add a subscription to a topic, broadcast to that topic, removing a subscription can all happen concurrently.

Reply:
thanks